com.unity.render-pipelines.universal/Shaders/LitExtractionPass.hlsl

```diff
@@ -0,0 +1,127 @@
+    inputData.normalizedScreenSpaceUV = input.positionCS.xy;
+}
+
+Varyings ExtractionVertex(Attributes input)
+{
+    Varyings output = (Varyings)0;
+
+    UNITY_SETUP_INSTANCE_ID(input);
+    UNITY_TRANSFER_INSTANCE_ID(input, output);
+    UNITY_INITIALIZE_VERTEX_OUTPUT_STEREO(output);
+
+    VertexPositionInputs vertexInput = GetVertexPositionInputs(input.positionOS.xyz);
+
+    // normalWS and tangentWS already normalize.
+    // this is required to avoid skewing the direction during interpolation
+    // also required for per-vertex lighting and SH evaluation
+    VertexNormalInputs normalInput = GetVertexNormalInputs(input.normalOS, input.tangentOS);
+
+    output.uv = TRANSFORM_TEX(input.texcoord, _BaseMap);
+
+    // already normalized from normal transform to WS.
+    output.normalWS = normalInput.normalWS;
+    real sign = input.tangentOS.w * GetOddNegativeScale();
+    output.tangentWS = half4(normalInput.tangentWS.xyz, sign);
+
+    output.positionWS = vertexInput.positionWS;
+    output.positionCS = vertexInput.positionCS;
+
+    if (UNITY_DataExtraction_Space == 0)
+        output.positionCS = float4(input.texcoord, 0.0F, 1.0f);
+    else if (UNITY_DataExtraction_Space == 1)
+        output.positionCS = float4(input.uv1.xy, 0.0F, 1.0f);
+    else if (UNITY_DataExtraction_Space == 2)
+        output.positionCS = float4(input.uv2, 0.0F, 1.0f);
+    else if (UNITY_DataExtraction_Space == 3)
+        output.positionCS = float4(input.uv3, 0.0F, 1.0f);
+    else if (UNITY_DataExtraction_Space == 4)
+        output.positionCS = float4(input.uv4, 0.0F, 1.0f);
+    else if (UNITY_DataExtraction_Space == 5)
+        output.positionCS = float4(input.uv5, 0.0F, 1.0f);
+    else if (UNITY_DataExtraction_Space == 6)
+        output.positionCS = float4(input.uv6, 0.0F, 1.0f);
+    else if (UNITY_DataExtraction_Space == 7)
+        output.positionCS = float4(input.uv7, 0.0F, 1.0f);
+
+    return output;
+}
+
+float4 ExtractionFragment(Varyings input) : SV_Target
+{
+    UNITY_SETUP_INSTANCE_ID(input);
+    UNITY_SETUP_STEREO_EYE_INDEX_POST_VERTEX(input);
+
+    SurfaceData surfaceData;
+    InitializeStandardLitSurfaceData(input.uv, surfaceData);
+
+    InputData inputData;
+    InitializeInputData(input, surfaceData.normalTS, inputData);
+
+    ExtractionInputs extraction;
+    extraction.vertexNormalWS = input.normalWS;
+    extraction.pixelNormalWS = inputData.normalWS;
+    extraction.positionWS = inputData.positionWS;
+    extraction.baseColor = surfaceData.albedo;
+    extraction.alpha = OutputAlpha(UniversalFragmentPBR(inputData, surfaceData).a);
+    #ifdef _SPECULAR_SETUP
+    extraction.specular = surfaceData.specular;
+    #else
+    extraction.metallic = surfaceData.metallic;
+    #endif
+    extraction.smoothness = surfaceData.smoothness;
+    extraction.occlusion = surfaceData.occlusion;
+    extraction.emission = surfaceData.emission;
+
+    return OutputExtraction(extraction);
```

Comment:
Won't this produce a warning if OutputExtraction() is declared as returning float4 (as it seems to be in DataExtraction.hlsl?

Reply:
Perhaps you mean the other callsite, where the return value seems to be a half4?

Reply:
Yes, you're quite right.  I reviewed some of this out of order and only remembered that OutputExtraction() was defined to be float4, but was returned from a function that was defined as half4.  The callsite I was concerned about is in  com.unity.render-pipelines.universal/Editor/ShaderGraph/Includes/PBRForwardPass.hlsl